src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -817,11 +817,25 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         } else if (ASR::is_a<ASR::List_t>(*left_type) && ASR::is_a<ASR::List_t>(*right_type)
                    && op == ASR::binopType::Add) {
             dest_type = left_type;
+            ASR::ttype_t *left_type2 = ASR::down_cast<ASR::List_t>(left_type)->m_type;
+            ASR::ttype_t *right_type2 = ASR::down_cast<ASR::List_t>(right_type)->m_type;
+            std::string ltype = ASRUtils::type_to_str_python(left_type2);
+            std::string rtype = ASRUtils::type_to_str_python(right_type2);
+            if (!ASRUtils::check_equal_type(left_type2, right_type2)) {
+                diag.add(diag::Diagnostic(
+                    "Both the lists should be of the same type for concatenation.",
+                    diag::Level::Error, diag::Stage::Semantic, {
+                        diag::Label("type mismatch (list[" + ltype + "] and list[" + rtype + "])",
+                                {left->base.loc, right->base.loc})
+                    })
+                );
+                throw SemanticAbort();
+            }
```

Comment:
We should upgrade `ASRUtils::type_to_str_python` to return `list[i32]` for this:
```suggestion
            std::string ltype = ASRUtils::type_to_str_python(left_type);
            std::string rtype = ASRUtils::type_to_str_python(right_type);
            if (!ASRUtils::check_equal_type(left_type, right_type)) {
                diag.add(diag::Diagnostic(
                    "Both the lists should be of the same type for concatenation.",
                    diag::Level::Error, diag::Stage::Semantic, {
                        diag::Label("type mismatch ('" + ltype + "' and '" + rtype + "')",
                                {left->base.loc, right->base.loc})
                    })
                );
                throw SemanticAbort();
            }
```
As well as `check_equal_type` to work for lists.